### ISSUE TITLE
No Chocolatey installation

### ISSUE THREAD

**Issue by fwinkelbauer** [chocolatey/boxstarter#369]
### What You Are Seeing?

The documentation of [Invoke-ChocolateyBoxstarter](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Invoke-ChocolateyBoxstarter.ps1) and [Install-BoxstarterPackage.ps1](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Install-BoxstarterPackage.ps1) state that Boxstarter will install Chocolatey on the machine if it does not exist. Using any of these functions on a fresh machine (using a local copy of Boxstarter) does not install `choco.exe`.

### What is Expected?

The above functions should either install Chocolatey, or the documentation should be updated.

### How Did You Get This To Happen? (Steps to Reproduce)

I am using [this repository](https://github.com/fwinkelbauer/windows-box-setup) to create a `box.zip` file which contains a copy of Boxstarter. Running `install-mybox.ps1` (included in `box.zip`) will install several Chocolatey packages on the machine (`C:\ProgramData\chocolatey\lib`), but there seems to be no `C:\ProgramData\chocolatey\bin\choco.exe`.

**Comment by mwallner** [comment]
Thanks for reporting, imo this is valid behaviour, or at least that what I expected it to be.

I'll update the documentation.
CC @pauby @flcdrg - any objections?

**Comment by mwallner** [comment]
I've just checked boxstarter internals,- and as far as I can see it always uses it's "own" version of Chocolatey.
Therefore it really doesn't make sense to automatically install Chocolatey when using Boxstarter, because it's not necessary.

**Comment by fwinkelbauer** [comment]
Installing Chocolatey would be useful if a user wants to update his packages later on. This case could be covered by adding `choco install chocolatey` to a Boxstarter script though.d

**Comment by fwinkelbauer** [comment]
Here's one thing to keep in mind:

Some of Chocolatey's features are not available if Boxstarter does not install a local version of Chocolatey. The `C:\ProgramData\Chocolatey\bin` folder is not added to PATH.

You can reproduce this on a fresh Windows machine (without Chocolatey) with this Boxstarter script (and a copy of Boxstarter, as described in my initial post above):

``` powershell
cinst bumpy.portable
Update-SessionEnvironment
bumpy
```

Which results in the error "bumpy not found", even though `C:\ProgramData\Chocolatey\bin\bumpy.exe` is installed.

Packages such as `vscode` seem to add their executable to PATH on their own, which is why something like this:

``` powershell
cinst vscode
Update-SessionEnvironment
code --install-extension cake-build.cake-vscode
```

works.

Here is a working snippet:

``` powershell
cinst chocolatey
cinst bumpy.portable
Update-SessionEnvironment
bumpy
```

**Comment by pauby** [comment]
@fwinkelbauer This would still be resolved by [installing Chocolatey manually](https://github.com/chocolatey/boxstarter/issues/369#issuecomment-441378507)?

**Comment by fwinkelbauer** [comment]
@pauby I have just updated the above comment. Yes, installing Chocolatey resolves this problem. But this behaviour might be confusing for some users, as this snippet:

``` powershell
cinst bumpy.portable
Update-SessionEnvironment
bumpy
```

will work on a machine with Chocolatey and Boxstarter installed, but it will fail when using a standalone version of Boxstarter.

**Comment by pauby** [comment]
@fwinkelbauer I agree. Maybe just an update to the docs in this case as @mwallner suggested?

**Comment by fwinkelbauer** [comment]
Which would be option number three, yes, that would work as well.

Another option would be that a user manually adds the Choco bin folder to `$env:Path` (as an alternative to `choco install chocolatey`) to his Boxstarter script.

**Comment by pauby** [comment]
> Another option would be that a user manually adds the Choco bin folder to $env:Path (as an alternative to choco install chocolatey) to his Boxstarter script.

That's for the user to decide though. It's not something we'd suggest as it's a kludge.

I'm not in favour of options 1 and 2 as that changes expected behaviour. What would be an option I think we should look at is Boxstarter installing Chocolatey itself for a future version (perhaps v3). @mwallner As you've looked at this above what are your thoughts?

**Comment by mwallner** [comment]
>  What would be an option I think we should look at is Boxstarter installing Chocolatey itself for a future version (perhaps v3).

Something we can consider yes - but we'd be dropping the possibility to install / "push install" packages without actually installing Chocolatey...

What about yet another option: just adding `C:\ProgramData\Chocolatey\bin` to `$env:Path` during Boxstarter operations?
- I think this would fix the issues @fwinkelbauer is currently facing.

I kind of don't want to drop the ability to remote-push Chocolatey packages via Boxstarter without forcing the remote computer to install Chocolatey itself.

CC @ferventcoder - if you've got a preference here, please let us know ;-)

**Comment by fwinkelbauer** [comment]
> What about yet another option: just adding `C:\ProgramData\Chocolatey\bin` to `$env:Path` during Boxstarter operations?
> 
>     * I think this would fix the issues @fwinkelbauer is currently facing.

That's pretty much what I was trying to say in my above comment. Adding the bin folder to `$env:Path` would "fix" every shim that Boxstarter would create via a Chocolatey installation.

**Comment by pauby** [comment]
@mwallner @fwinkelbauer 

> What about yet another option: just adding C:\ProgramData\Chocolatey\bin to $env:Path during Boxstarter operations?

Do we know why that's not being done? I'd rather understand that before we just go ahead and fix it.

**Comment by fwinkelbauer** [comment]
My understanding is that this is not happening because Chocolatey is not installed when using the standalone version of Boxstarter.

[Installing Boxstarter](https://boxstarter.org/InstallBoxstarter) using any other method includes an installation of Chocolatey:

# 1. Using Chocolatey

`choco install boxstarter`. This should be obvious, Chocolatey is already installed.

# 2. Using the ZIP Archive provided on GitHub

This approach uses the Boxstarter bootstrapper, see below.

# 3. Using the Boxstarter Bootstrapper

The [Boxstarter bootstrapper](https://boxstarter.org/bootstrapper.ps1) ensures that Chocolatey is installed.

# 4. ClickOnce

The [ClickOnce launcher](https://github.com/chocolatey/boxstarter/blob/4c8c3500b8676755eba054530dbb33af4b9a6d3d/Boxstarter.ClickOnce/Launcher.cs#L36) seems to utilize the `Boxstarter.bat` file, which should be the same method used in 2. (ZIP Archive).

**Comment by fwinkelbauer** [comment]
I'll try to summarize what we already know, maybe this will help to clear things up.

Any method described [here](https://boxstarter.org/InstallBoxstarter) which installs Boxstarter, installs Chocolatey as part of the "bootstrapping" (see my post above).

The Powershell doc for [Invoke-ChocolateyBoxstarter](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Invoke-ChocolateyBoxstarter.ps1) and [Install-BoxstarterPackage](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Install-BoxstarterPackage.ps1) claim that they will ensure Chocolatey is installed - which is currently not true. These installation checks are only done when using the bootstrapper.

The section "Running an Install from a network share or external media" [here](https://boxstarter.org/InstallingPackages) describes a method to install packages on a Computer without bootstrapping Boxstarter or Chocolatey. This is the only way (that I am aware of) in which Chocolatey is not installed on a computer. Boxstarter uses its own copy of Chocolatey to install packages. This approach works fine for package installation, but creates problems later on because any shim that Chocolatey creates is not available on `$env:Path` (because no one added `C:\ProgramData\Chocolatey\bin` to `$env:Path`).

In short:

- The documentation for [Invoke-ChocolateyBoxstarter](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Invoke-ChocolateyBoxstarter.ps1) and [Install-BoxstarterPackage](https://github.com/chocolatey/boxstarter/blob/master/Boxstarter.Chocolatey/Install-BoxstarterPackage.ps1) is outdated
- Using a copy of Boxstarter without any bootstrapping mechanism does not install Chocolatey on a machine. As a result `C:\ProgramData\Chocolatey\bin` is not added to `$env:Path`.

**Comment by pauby** [comment]
My thoughts on this are to update the documentation to make it clear that Chocolatey is NOT installed by `Invoke-ChocolateyBoxstarter` or `Install-BoxstarterPackage`. And to make it clear to the user that if they want the additional functionality offered by Chocolatey that they specifically install it.

I'm assuming that `c:\programdata\chocolatey\bin` doesn't exist as Chocolatey has not been installed so adding it to the path requires it to be created. I will need to check if it's actually there!

/cc @mwallner @flcdrg
